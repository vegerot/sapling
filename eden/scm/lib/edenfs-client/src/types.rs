/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Convert between Thrift and native Rust types.

use serde::Serialize;
use thrift_types::edenfs;
pub use types::workingcopy_client::CheckoutConflict;
pub use types::workingcopy_client::CheckoutMode;
pub use types::workingcopy_client::ConflictType;
pub use types::workingcopy_client::FileStatus;
use types::RepoPathBuf;

// edenfs::EdenError
#[derive(Debug, Serialize)]
pub struct EdenError {
    pub message: String,
    pub error_code: Option<i32>,
    pub error_type: String,
}

/// Crate-local `From` to workaround Rust orphan rule.
pub trait LocalFrom<T> {
    fn local_from(v: T) -> Self;
}
pub trait LocalTryFrom<T>: Sized {
    type Error;
    fn local_try_from(v: T) -> Result<Self, Self::Error>;
}

impl LocalFrom<edenfs::ScmFileStatus> for FileStatus {
    fn local_from(status: edenfs::ScmFileStatus) -> Self {
        match status {
            edenfs::ScmFileStatus::ADDED => FileStatus::Added,
            edenfs::ScmFileStatus::MODIFIED => FileStatus::Modified,
            edenfs::ScmFileStatus::REMOVED => FileStatus::Removed,
            edenfs::ScmFileStatus::IGNORED => FileStatus::Ignored,
            _ => panic!("unexpected ScmFileStatus: {}", status),
        }
    }
}

impl LocalFrom<CheckoutMode> for edenfs::CheckoutMode {
    fn local_from(val: CheckoutMode) -> Self {
        match val {
            CheckoutMode::Normal => edenfs::CheckoutMode::NORMAL,
            CheckoutMode::Force => edenfs::CheckoutMode::FORCE,
            CheckoutMode::DryRun => edenfs::CheckoutMode::DRY_RUN,
        }
    }
}

impl LocalFrom<edenfs::ConflictType> for ConflictType {
    fn local_from(conflict_type: edenfs::ConflictType) -> Self {
        match conflict_type {
            edenfs::ConflictType::ERROR => ConflictType::Error,
            edenfs::ConflictType::MODIFIED_REMOVED => ConflictType::ModifiedRemoved,
            edenfs::ConflictType::UNTRACKED_ADDED => ConflictType::UntrackedAdded,
            edenfs::ConflictType::REMOVED_MODIFIED => ConflictType::RemovedModified,
            edenfs::ConflictType::MISSING_REMOVED => ConflictType::MissingRemoved,
            edenfs::ConflictType::MODIFIED_MODIFIED => ConflictType::ModifiedModified,
            edenfs::ConflictType::DIRECTORY_NOT_EMPTY => ConflictType::DirectoryNotEmpty,
            _ => panic!("unexpected ConflictType: {}", conflict_type),
        }
    }
}

impl LocalTryFrom<edenfs::CheckoutConflict> for CheckoutConflict {
    type Error = ();

    fn local_try_from(conflict: edenfs::CheckoutConflict) -> Result<Self, Self::Error> {
        let path = RepoPathBuf::from_utf8(conflict.path).map_err(|e| {
            tracing::warn!("conflict: ignore non-utf8 path {}", e);
        })?;
        let conflict_type = ConflictType::local_from(conflict.r#type);
        let message = conflict.message;
        Ok(CheckoutConflict {
            path,
            conflict_type,
            message,
        })
    }
}

impl std::fmt::Display for EdenError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "EdenError: {}", self.message)
    }
}

impl std::error::Error for EdenError {}

impl LocalTryFrom<&(dyn std::error::Error + 'static)> for EdenError {
    type Error = ();

    fn local_try_from(value: &(dyn std::error::Error + 'static)) -> Result<Self, Self::Error> {
        if let Some(value) = value.downcast_ref::<edenfs::EdenError>() {
            let eden_error = Self {
                message: value.message.clone(),
                error_code: value.errorCode,
                error_type: value.errorType.to_string(),
            };
            Ok(eden_error)
        } else {
            Err(())
        }
    }
}
