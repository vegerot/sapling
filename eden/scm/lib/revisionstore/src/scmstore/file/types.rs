/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod auxdata;
mod lazy_file;
mod store_file;

pub use self::attrs::FileAttributes;
pub use self::auxdata::FileAuxData;
pub(crate) use self::lazy_file::LazyFile;
pub use self::store_file::StoreFile;
