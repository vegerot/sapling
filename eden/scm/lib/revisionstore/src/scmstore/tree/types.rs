/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod lazy_tree;
mod store_tree;

pub use self::attrs::TreeAttributes;
pub(crate) use self::lazy_tree::AuxData;
pub(crate) use self::lazy_tree::LazyTree;
pub use self::store_tree::StoreTree;
