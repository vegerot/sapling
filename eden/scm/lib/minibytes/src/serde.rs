/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::cell::RefCell;
use std::fmt;

use serde::de;
use serde::Deserialize;
use serde::Deserializer;
use serde::Serialize;
use serde::Serializer;

use crate::Bytes;
use crate::Text;

impl Serialize for Bytes {
    #[inline]
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_bytes(self)
    }
}

impl Serialize for Text {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(self)
    }
}

struct BytesVisitor;
struct TextVisitor;

thread_local! {
    static DESERIALIZE_HINT: RefCell<Option<Bytes>> = const { RefCell::new(None) };
}

fn set_deserialize_hint(bytes: Option<Bytes>) -> Option<Bytes> {
    DESERIALIZE_HINT.with_borrow_mut(|f| {
        let orig = f.take();
        *f = bytes;
        orig
    })
}

impl Bytes {
    /// Call `func` with a "deserialize hint" as an attempt to avoid `memcpy`s.
    /// `func` is usually a serde deserialize function taking `self` as input.
    ///
    /// Only affects the current thread, with the assumption that serde
    /// deserialize is usually single threaded.
    pub fn as_deserialize_hint<R>(&self, func: impl Fn() -> R) -> R {
        let orig = set_deserialize_hint(Some(self.clone()));
        let result = func();
        set_deserialize_hint(orig);
        result
    }
}

impl<'de> de::Visitor<'de> for BytesVisitor {
    type Value = Bytes;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("byte slice")
    }

    fn visit_borrowed_bytes<E: de::Error>(self, v: &'de [u8]) -> Result<Self::Value, E> {
        self.visit_bytes(v)
    }

    fn visit_borrowed_str<E: de::Error>(self, v: &'de str) -> Result<Self::Value, E> {
        Ok(Bytes::copy_from_slice(v.as_bytes()))
    }

    fn visit_byte_buf<E>(self, v: Vec<u8>) -> Result<Self::Value, E> {
        Ok(Bytes::from(v))
    }

    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E> {
        let bytes = DESERIALIZE_HINT.with_borrow(|parent_buffer| match parent_buffer {
            Some(buf) => buf.slice_to_bytes(v),
            None => Bytes::copy_from_slice(v),
        });
        Ok(bytes)
    }
}

impl<'de> Deserialize<'de> for Bytes {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        deserializer.deserialize_byte_buf(BytesVisitor)
    }
}

impl<'de> de::Visitor<'de> for TextVisitor {
    type Value = Text;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("str")
    }

    fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
        let text = DESERIALIZE_HINT.with_borrow(|parent_buffer| match parent_buffer {
            Some(buf) => {
                let buf = buf.slice_to_bytes(v.as_bytes());
                // safety: `v: &str` proves `v` and therefore `buf` valid utf-8.
                unsafe { Text::from_utf8_unchecked(buf) }
            }
            None => Text::copy_from_slice(v),
        });
        Ok(text)
    }

    fn visit_string<E: de::Error>(self, v: String) -> Result<Self::Value, E> {
        Ok(Text::from(v))
    }
}

impl<'de> Deserialize<'de> for Text {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        deserializer.deserialize_string(TextVisitor)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Deserialize, Serialize)]
    struct S {
        a: Bytes,
        b: Bytes,
        c: Text,
    }

    #[test]
    fn test_deserialize_hint() {
        let s1 = S {
            a: Bytes::copy_from_slice(b"aaaa"),
            b: Bytes::from_static(b"bbbb"),
            c: Text::from_static("ccc"),
        };
        let serialized = Bytes::from(serde_cbor::to_vec(&s1).unwrap());

        // Deserialize directly - no zero copy.
        let s2: S = serde_cbor::from_slice(&serialized).unwrap();
        assert!(serialized.range_of_slice(s2.a.as_ref()).is_none());
        assert!(serialized.range_of_slice(s2.b.as_ref()).is_none());
        assert!(serialized.range_of_slice(s2.c.as_bytes()).is_none());

        // Deserialize with hint - can be zero copy.
        let s3: S = serialized.as_deserialize_hint(|| serde_cbor::from_slice(&serialized).unwrap());
        assert!(serialized.range_of_slice(s3.a.as_ref()).is_some());
        assert!(serialized.range_of_slice(s3.b.as_ref()).is_some());
        assert!(serialized.range_of_slice(s3.c.as_bytes()).is_some());
    }
}
