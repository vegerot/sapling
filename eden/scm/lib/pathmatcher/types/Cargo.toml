# @generated by autocargo from //eden/scm/lib/pathmatcher/types:pathmatcher-types

[package]
name = "pathmatcher-types"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.86"
types = { version = "0.1.0", path = "../../types" }
