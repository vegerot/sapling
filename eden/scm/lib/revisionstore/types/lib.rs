/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! revisionstore_types - Data types used by the revisionstore crate

mod datastore;

pub use crate::datastore::Metadata;
