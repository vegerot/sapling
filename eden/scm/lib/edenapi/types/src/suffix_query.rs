/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#[cfg(any(test, feature = "for-tests"))]
use quickcheck_arbitrary_derive::Arbitrary;
use serde::Deserialize;
use serde::Serialize;
use type_macros::auto_wire;
use types::RepoPathBuf;

use crate::CommitId;

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct SuffixQueryRequest {
    #[id(0)]
    pub commit: CommitId,
    #[id(1)]
    pub basename_suffixes: Vec<String>,
    #[id(2)]
    pub prefixes: Option<Vec<String>>,
}

#[auto_wire]
#[derive(Clone, Default, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[cfg_attr(any(test, feature = "for-tests"), derive(Arbitrary))]
pub struct SuffixQueryResponse {
    #[id(0)]
    pub file_path: RepoPathBuf,
}
