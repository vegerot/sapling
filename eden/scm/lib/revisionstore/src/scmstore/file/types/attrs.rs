/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::ops::BitAnd;
use std::ops::BitOr;
use std::ops::BitOrAssign;
use std::ops::Not;
use std::ops::Sub;

use edenapi_types::FileAttributes as SaplingRemoteApiFileAttributes;
use serde::Deserialize;
use serde::Serialize;

use crate::scmstore::attrs::StoreAttrs;

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct FileAttributes {
    pub content: bool,
    pub aux_data: bool,
}

impl From<FileAttributes> for SaplingRemoteApiFileAttributes {
    fn from(v: FileAttributes) -> Self {
        SaplingRemoteApiFileAttributes {
            content: v.content,
            aux_data: v.aux_data,
        }
    }
}

impl StoreAttrs for FileAttributes {
    const NONE: Self = FileAttributes {
        content: false,
        aux_data: false,
    };

    /// Returns all the attributes which are present or can be computed from present attributes.
    fn with_computable(&self) -> FileAttributes {
        if self.content {
            *self | FileAttributes::AUX
        } else {
            *self
        }
    }
}

impl FileAttributes {
    pub const CONTENT: Self = FileAttributes {
        content: true,
        aux_data: false,
    };

    pub const AUX: Self = FileAttributes {
        content: false,
        aux_data: true,
    };
}

impl Not for FileAttributes {
    type Output = Self;

    fn not(self) -> Self::Output {
        FileAttributes {
            content: !self.content,
            aux_data: !self.aux_data,
        }
    }
}

impl BitAnd for FileAttributes {
    type Output = Self;

    fn bitand(self, rhs: Self) -> Self::Output {
        FileAttributes {
            content: self.content & rhs.content,
            aux_data: self.aux_data & rhs.aux_data,
        }
    }
}

impl BitOr for FileAttributes {
    type Output = Self;

    fn bitor(self, rhs: Self) -> Self::Output {
        FileAttributes {
            content: self.content | rhs.content,
            aux_data: self.aux_data | rhs.aux_data,
        }
    }
}

impl BitOrAssign for FileAttributes {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

/// The subtraction operator is implemented here to mean "set difference" aka relative complement.
impl Sub for FileAttributes {
    type Output = Self;

    fn sub(self, rhs: Self) -> Self::Output {
        self & !rhs
    }
}
