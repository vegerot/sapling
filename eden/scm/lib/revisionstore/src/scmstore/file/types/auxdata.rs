/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use edenapi_types::Blake3;
use edenapi_types::ContentId;
use edenapi_types::FileAuxData as EdenApiFileAuxData;
use edenapi_types::Sha1;
use serde::Deserialize;
use serde::Serialize;
use types::Sha256;

use crate::indexedlogauxstore::Entry as AuxDataEntry;

#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct FileAuxData {
    pub total_size: u64,
    pub content_id: ContentId,
    pub content_sha1: Sha1,
    pub content_sha256: Sha256,
    pub content_seeded_blake3: Option<Blake3>,
}

impl From<AuxDataEntry> for FileAuxData {
    fn from(v: AuxDataEntry) -> Self {
        FileAuxData {
            total_size: v.total_size(),
            content_id: v.content_id(),
            content_sha1: v.content_sha1(),
            content_sha256: Sha256::from_byte_array(v.content_sha256().into()),
            content_seeded_blake3: v.content_seeded_blake3(),
        }
    }
}

impl From<FileAuxData> for AuxDataEntry {
    fn from(v: FileAuxData) -> Self {
        AuxDataEntry {
            total_size: v.total_size,
            content_id: v.content_id,
            content_sha1: v.content_sha1,
            content_sha256: v.content_sha256.into_inner().into(),
            content_seeded_blake3: v.content_seeded_blake3,
        }
    }
}

impl From<EdenApiFileAuxData> for FileAuxData {
    fn from(v: EdenApiFileAuxData) -> Self {
        FileAuxData {
            total_size: v.total_size,
            content_id: v.content_id,
            content_sha1: v.sha1,
            content_sha256: Sha256::from_byte_array(v.sha256.into()),
            content_seeded_blake3: v.seeded_blake3,
        }
    }
}

impl From<FileAuxData> for EdenApiFileAuxData {
    fn from(v: FileAuxData) -> Self {
        EdenApiFileAuxData {
            total_size: v.total_size,
            content_id: v.content_id,
            sha1: v.content_sha1,
            sha256: v.content_sha256.into_inner().into(),
            seeded_blake3: v.content_seeded_blake3,
        }
    }
}
