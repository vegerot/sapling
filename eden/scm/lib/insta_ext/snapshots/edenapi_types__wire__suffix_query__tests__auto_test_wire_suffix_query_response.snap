---
source: suffix_query.rs
expression: "WireSuffixQueryResponse::arbitrary(&mut g)"
---
{"0":"jaiffe/nuat/fk/ejkleav/rc/ohh"}
